tpu-client-next/src/node_address_service/leader_tpu_cache_service.rs

```diff
@@ -0,0 +1,520 @@
+    async fn slot_leaders(
+        &self,
+        estimated_current_slot: Slot,
+        slots_in_epoch: Slot,
+    ) -> Result<Vec<Pubkey>, Error>;
+}
+
+async fn update_leader_info(
+    estimated_current_slot: Slot,
+    rpc_client: &impl ClusterInfoProvider,
+    epoch_info: &mut EpochInfo,
+    slot_leaders: &mut SlotLeaders,
+    num_consecutive_failures: &mut usize,
+    max_consecutive_failures: usize,
+) -> Result<(), Error> {
+    if estimated_current_slot > epoch_info.last_slot_in_epoch {
+        try_update(
+            "epoch info",
+            epoch_info,
+            || EpochInfo::new(rpc_client, estimated_current_slot),
+            num_consecutive_failures,
+            max_consecutive_failures,
+        )
+        .await?;
+    }
+    if estimated_current_slot.saturating_add(MAX_FANOUT_SLOTS) > slot_leaders.last_slot() {
+        try_update(
+            "slot leaders",
+            slot_leaders,
+            || {
+                SlotLeaders::new(
+                    rpc_client,
+                    estimated_current_slot,
+                    epoch_info.slots_in_epoch,
+                )
+            },
+            num_consecutive_failures,
+            max_consecutive_failures,
+        )
+        .await?;
+    }
+    Ok(())
+}
+
+/// Get the TPU sockets for the current and upcoming leaders according to
+/// fanout size.
+fn leader_sockets(
+    estimated_current_slot: Slot,
+    lookahead_leaders: u8,
+    slot_leaders: &SlotLeaders,
+    leader_tpu_map: &LeaderTpuMap,
+) -> Vec<SocketAddr> {
+    let fanout_slots = (lookahead_leaders as u64).saturating_mul(NUM_CONSECUTIVE_LEADER_SLOTS);
+    let mut leader_sockets = Vec::with_capacity(lookahead_leaders as usize);
+    // `first_slot` might have been advanced since caller last read the
+    // `estimated_current_slot` value. Take the greater of the two values to
+    // ensure we are reading from the latest leader schedule.
+    let current_slot = std::cmp::max(estimated_current_slot, slot_leaders.first_slot);
+    for leader_slot in
+        (current_slot..current_slot + fanout_slots).step_by(NUM_CONSECUTIVE_LEADER_SLOTS as usize)
+    {
+        if let Some(leader) = slot_leaders.slot_leader(leader_slot) {
+            if let Some(tpu_socket) = leader_tpu_map.get(leader) {
+                leader_sockets.push(*tpu_socket);
+                debug!("Pushed leader {leader} TPU socket: {tpu_socket}");
+            } else {
+                // The leader is probably delinquent
+                debug!("TPU not available for leader {leader}");
+            }
+        } else {
+            // Overran the local leader schedule cache
+            warn!(
+                "Leader not known for slot {}; cache holds slots [{},{}]",
+                leader_slot,
+                slot_leaders.first_slot,
+                slot_leaders.last_slot()
+            );
+        }
+    }
+
+    leader_sockets
```

Comment:
might want to comment how the caller is supposed to deal with empty Vec

Reply:
I added a comment to the function itself.